src/pcp/types.rs

```diff
@@ -74,6 +74,107 @@ impl<F: FieldElement> Value<F, Mul<F>> for Boolean<F> {
     }
 }
 
+/// This type represents vectors for which `poly(x) == 0` holds for some polynomial `poly` and each
+/// vector element `x`.
+///
+/// This type is "generic" in that it can be used to construct high level types. For example, a
+/// boolean vector can be represented as follows:
+///
+/// ```
+/// use prio::field::{Field64, FieldElement};
+/// use prio::pcp::types::PolyCheckedVector;
+///
+/// let data = vec![Field64::zero(), Field64::one(), Field64::zero()];
+/// let x = PolyCheckedVector::new_range_checked(data, 0, 2);
+/// ```
+///
+/// The proof size for this data type is `O(sqrt(d * m))`, where `m` is the length of the input
+/// vector and `d` is the degree of the polynomial. (For the bit vector example above, `d=2`.) The
+/// validity circuit is based on the construction of \[[BBG+19](https://eprint.iacr.org/2019/188),
+/// Corollary 4.9\].
+pub struct PolyCheckedVector<F: FieldElement> {
+    data: Vec<F>,
+    poly: Vec<F>,
+    chunk_len: usize,
+}
+
+impl<F: FieldElement> PolyCheckedVector<F> {
+    /// Returns a `poly`-checked vector where `poly(x) == 0` if and only if `x` is in range
+    /// `[start, end)`. The degree of `poly` is equal to `end`.
+    pub fn new_range_checked(data: Vec<F>, start: usize, end: usize) -> Self {
+        let chunk_len = (data.len() as f64).sqrt() as usize;
+        Self {
+            data,
+            poly: poly_range_check(start, end),
+            chunk_len,
+        }
+    }
+}
+
+impl<F: FieldElement> Value<F, ParallelSum<F, PolyEval<F>>> for PolyCheckedVector<F> {
+    fn valid(&self, g: &mut dyn GadgetCallOnly<F>, rand: &[F]) -> Result<F, PcpError> {
+        if rand.len() < 1 {
+            return Err(PcpError::CircuitIn("random input not long enough"));
+        }
+
+        let mut r = rand[0];
+        let mut outp = F::zero();
+        for chunk in self.data.chunks(self.chunk_len) {
+            let n = chunk.len();
+            let v = match n == self.chunk_len {
```

Comment:
This works fine if you want to keep it but you could make this a little less indented if you do 
```
let v = if n == self.chunk_len {
    g.call(chunk)?
} else {
    // the other thing
};
```

Reply:
Done.